style: Basic @import support.

emilio opened the pull request:
r? @SimonSapin or @mbrubeck 

cc @heycam and @bholley

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14540)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/media_queries.rs, components/style/values/specified/url.rs, components/style/stylesheets.rs, components/style/stylist.rs
 * @fitzgen: components/script/dom/cssrule.rs, components/script/lib.rs, components/script/dom/create.rs, components/script/dom/cssimportrule.rs, components/script/dom/htmlstyleelement.rs, components/script/dom/htmllinkelement.rs, components/script/dom/mod.rs, components/script/dom/webidls/CSSImportRule.webidl, components/script/stylesheet_loader.rs
 * @kichjang: components/script/dom/cssrule.rs, components/script/lib.rs, components/script/dom/create.rs, components/script/dom/cssimportrule.rs, components/script/dom/htmlstyleelement.rs, components/script/dom/htmllinkelement.rs, components/script/dom/mod.rs, components/script/dom/webidls/CSSImportRule.webidl, components/script/stylesheet_loader.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style and script code, but no tests are modified. Please consider adding a test!

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 40fe58d with merge 4697307...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4633)

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit fa62995 with merge 43f8679...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/4658)

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 39e3c9d with merge 6857f9e...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/4659)

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 18b0ffe has been approved by `emilio`

<!-- @bors-servo r=emilio 18b0ffe93b2e31dc722c0a1fae686f896ac63883 -->

emilio commented:
@bors-servo r- try

bors-servo commented:
:hourglass: Trying commit 18b0ffe with merge 2f65305...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/4661)

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit e6dd35e with merge fe040c8...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/4666)

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 59b1761 with merge 5e3d583...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4643)

emilio commented:
@bors-servo retry

 * infra

bors-servo commented:
:hourglass: Trying commit 59b1761 with merge 037371d...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1366)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14557) made this pull request unmergeable. Please resolve the merge conflicts.

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 9bb98fb with merge b04161a...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1413)

emilio commented:
@bors-servo try

 * infra

bors-servo commented:
:hourglass: Trying commit 65cfb51 with merge 3ff1a1b...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/4700)

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit a2d0684 with merge 34bb2d0...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/4701)

emilio commented:
@bors-servo try

 * I don't know how to rebase, sigh.

bors-servo commented:
:hourglass: Trying commit e44d5c8 with merge 2732867...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4677)

stshine commented:
@bors-servo try

emilio commented:
@bors-servo retry

 * I just want a decent list of test expectations to update :(

bors-servo commented:
:hourglass: Trying commit e44d5c8 with merge 4a6e12d...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/4703)

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit c08219b with merge e0d4116...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1393)

emilio commented on components/script/dom/cssrule.rs:
```diff
@@ -75,6 +76,7 @@ impl CSSRule {
                         rule: StyleCssRule) -> Root<CSSRule> {
         // be sure to update the match in as_specific when this is updated
         match rule {
+            StyleCssRule::Import(s) => Root::upcast(CSSImportRule::new(window, parent_stylesheet, s)),
```
Whoops, right, done.

larsbergstrom commented:
@bors-servo retry
- https://github.com/servo/saltfs/issues/553

bors-servo commented:
:hourglass: Trying commit c08219b with merge 9761dc5...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4688)

KiChjang commented:
@bors-servo retry infra

bors-servo commented:
:hourglass: Trying commit c08219b with merge d7bf4dc...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/4715)

emilio commented:
@bors-servo try 

 * Manifest changed? what?

bors-servo commented:
:hourglass: Trying commit 6b26edf with merge b63d7b3...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1449)

emilio commented:
@bors-servo retry 

 infra :'(

bors-servo commented:
:hourglass: Trying commit 6b26edf with merge 3a9dca4...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1416)

KiChjang commented:
```
Error running mach:

    ['test-css', '--release', '--processes', '16', '--log-raw', 'test-css.log', '--log-errorsummary', 'css-errorsummary.log', '--always-succeed']

The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.

You should consider filing a bug for this issue.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

AttributeError: 'NoneType' object has no attribute 'endswith'

  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 603, in test_css
    ret = self.run_test_list_or_dispatch(kwargs["test_list"], "css", self._test_css, **kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 434, in run_test_list_or_dispatch
    return correct_function(**kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 611, in _test_css
    return self.wptrunner(run_file, **kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 449, in wptrunner
    self.set_software_rendering_env(kwargs['release'])
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 738, in set_software_rendering_env
    set_osmesa_env(args[0], os.environ)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/command_base.py", line 219, in set_osmesa_env
    osmesa_path = path.join(find_dep_path_newest('osmesa-src', bin_path), "out", "lib", "gallium")
  File "/usr/lib/python2.7/posixpath.py", line 77, in join
    elif path == '' or path.endswith('/'):
```

emilio commented:
Funny situation, I can't run the tests locally because of https://github.com/servo/osmesa-src/issues/14, and it seems bors doesn't like to run tests on this pr... :(

I'll try to fix osmesa, I guess... meanwhile:

@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 6b26edf with merge 8d9bb3a...

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit a75f6a4 with merge 7f1ad03...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1418)

KiChjang commented:
Looks like we got good news!

<details>
```
Ran 9977 tests finished in 834.0 seconds.
  • 9925 ran as expected. 146 tests skipped.
  • 2 tests failed unexpectedly
  • 46 tests passed unexpectedly
  • 4 tests had unexpected subtest results

Tests with unexpected results:
  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-001.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-009.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-007.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-008.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-014.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-012.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-016.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-015.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-025.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-020.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-019.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-024.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-026.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-028.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-021.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-027.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-029.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-030.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-040.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-041.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-043.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-042.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-047.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-048.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-044.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-045.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-046.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-053.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-charset-060.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-import-002.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-import-003.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-import-005.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-import-007.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-import-006.htm

  ▶ FAIL [expected PASS] /css21_dev/html4/at-import-004.htm
  └   → /css21_dev/html4/at-import-004.htm 628af9911472946c94b9f3bbd65e9b3f51d641d6
/css21_dev/html4/reference/no-red-filler-text-ref.htm 8af7223716bda456f2e130bceda4e35ed7958dc8
Testing 628af9911472946c94b9f3bbd65e9b3f51d641d6 == 8af7223716bda456f2e130bceda4e35ed7958dc8

  ▶ PASS [expected FAIL] /css21_dev/html4/at-rule-005.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-keywords-000.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/at-rule-010.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/c11-import-000.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/cascade-012.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/core-syntax-009.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/eof-005.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/eof-006.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/media-dependency-002.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/media-dependency-004.htm

  ▶ PASS [expected FAIL] /css21_dev/html4/media-dependency-005.htm

  ▶ FAIL [expected PASS] /css21_dev/html4/media-dependency-008.htm
  └   → /css21_dev/html4/media-dependency-008.htm 628af9911472946c94b9f3bbd65e9b3f51d641d6
/css21_dev/html4/reference/no-red-filler-text-ref.htm 8af7223716bda456f2e130bceda4e35ed7958dc8
Testing 628af9911472946c94b9f3bbd65e9b3f51d641d6 == 8af7223716bda456f2e130bceda4e35ed7958dc8

  ▶ PASS [expected FAIL] /css21_dev/html4/uri-001.htm

  ▶ Unexpected subtest result in /cssom-1_dev/html/cssimportrule.htm:
  └ PASS [expected FAIL] CSSRule and CSSImportRule types

  ▶ Unexpected subtest result in /cssom-1_dev/html/cssimportrule.htm:
  └ PASS [expected FAIL] Type of CSSRule#type and constant values

  ▶ Unexpected subtest result in /css-transitions-1_dev/html/changing-while-transition.htm:
  └ PASS [expected FAIL] changing transition-duration / values

  ▶ Unexpected subtest result in /cssom-1_dev/html/index-003.htm:
  └ PASS [expected FAIL] @import rule is expected to be @import url("main.css")

  ▶ Unexpected subtest result in /cssom-1_dev/html/interfaces.htm:
  └ PASS [expected FAIL] CSSImportRule interface: existence and properties of interface object

  ▶ Unexpected subtest result in /cssom-1_dev/html/interfaces.htm:
  └ PASS [expected FAIL] CSSImportRule interface object length

  ▶ Unexpected subtest result in /cssom-1_dev/html/interfaces.htm:
  └ PASS [expected FAIL] CSSImportRule interface: existence and properties of interface prototype object

  ▶ Unexpected subtest result in /cssom-1_dev/html/interfaces.htm:
  └ PASS [expected FAIL] CSSImportRule interface: existence and properties of interface prototype object's "constructor" property
```
</details>

emilio commented:
I ended up being able to test locally with #14584. Just pushed the results, this should be ready for review.

r? @SimonSapin or @heycam 

@bors-servo try

bors-servo commented:
:hourglass: Trying commit f3acba7 with merge 5288751...

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 2fc060b with merge e6b2197...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1427)

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit a62d681 with merge ee715fb...

SimonSapin commented:
Reviewed 23 of 23 files at r5, 3 of 3 files at r6, 4 of 4 files at r7, 19 of 19 files at r8, 40 of 40 files at r9, 1 of 1 files at r10, 11 of 11 files at r11, 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[components/script/stylesheet_loader.rs, line 108 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ047pkT1ceNgNBWY2x:-KZ047pkT1ceNgNBWY2y:b-d80tjz) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/script/stylesheet_loader.rs#L108)):*
> ```Rust
>                                                         Some(environment_encoding),
>                                                         Origin::Author,
>                                                         media.clone(),
> ```

Replace this clone by the "option dance" like this code used to do when it was in `components/script/dom/htmllinkelement.rs`: make the field an `Option<MediaList>` and use `.take().unwrap()` here.

---

*[components/script/stylesheet_loader.rs, line 151 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ03B-Fe6JkRo4L-hhV:-KZ03B-G2PUlmgnISHrL:bjdwoh6) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/script/stylesheet_loader.rs#L151)):*
> ```Rust
>             StylesheetContextSource::Import(ref import) => {
>                 let import = import.read();
>                 import.url.url().expect("Tried to load an invalid url").clone()
> ```

This should not panic. Instead, an invalid URL should produce a similar result as a network error.

---

*[components/script/stylesheet_loader.rs, line 183 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ03SPfsnjxOQZ94Keu:-KZ03SPfsnjxOQZ94Kev:b-okuktc) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/script/stylesheet_loader.rs#L183)):*
> ```Rust
>             StylesheetContextSource::Import(ref import) => {
>                 let import = import.read();
>                 import.url.url().expect("Requested import with invalid url!").clone()
> ```

This should not panic. Instead, an invalid URL should produce a similar result as a network error.

---

*[components/style/media_queries.rs, line 20 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ04vsn4pSfIBu1A7XS:-KZ04vsn4pSfIBu1A7XT:b-ts0ptk) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/style/media_queries.rs#L20)):*
> ```Rust
> 
> 
> #[derive(Debug, PartialEq, Clone)]
> ```

Please remove this `Clone` impl, see my "option dance" comment in `stylesheet_loader.rs`.

---

*[components/style/stylesheets.rs, line 48 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ04ju4Invm5zspxip8:-KZ04ju4Invm5zspxip9:b51i6pr) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/style/stylesheets.rs#L48)):*
> ```Rust
> }
> 
> #[derive(Default, Debug, Clone)]
> ```

Please revert this, it is never used.

---

*[components/style/stylist.rs, line 101 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ-qxdrCVWIfOI36wv7:-KZ-qxdrCVWIfOI36wv8:b-51ydme) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/style/stylist.rs#L101)):*
> ```Rust
> }
> 
> // I know...
> ```

This comment is not particularly helpful. I’d rather have it explain why this function takes some many arguments, but we can do better and fix that.

Please apply the diff below and move the function back to a `Stylist` method that takes `&mut self`. The new `Arc` does add some runtime cost, but I think it’s small enough that cleaner code is worth it.

```diff
diff --git a/components/style/stylist.rs b/components/style/stylist.rs
index 5b40dcf..5abef17 100644
--- a/components/style/stylist.rs
+++ b/components/style/stylist.rs
@@ -49,7 +49,7 @@ pub type FnvHashMap<K, V> = HashMap<K, V, BuildHasherDefault<::fnv::FnvHasher>>;
 #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
 pub struct Stylist {
     /// Device that the stylist is currently evaluating against.
-    pub device: Device,
+    pub device: Arc<Device>,
 
     /// Viewport constraints based on the current device.
     viewport_constraints: Option<ViewportConstraints>,
@@ -98,83 +98,64 @@ pub struct Stylist {
     non_common_style_affecting_attributes_selectors: Vec<Selector<SelectorImpl>>,
 }
 
-// I know...
-fn add_stylesheet_internal(
-    stylesheet: &Stylesheet,
-    device: &Device,
-    element_map: &mut PerPseudoElementSelectorMap,
-    pseudos_map: &mut FnvHashMap<PseudoElement, PerPseudoElementSelectorMap>,
-    animations: &mut FnvHashMap<Atom, KeyframesAnimation>,
-    precomputed_pseudo_element_decls: &mut FnvHashMap<PseudoElement, Vec<ApplicableDeclarationBlock>>,
-    rules_source_order: &mut usize,
-    state_deps: &mut DependencySet,
-    sibling_affecting_selectors: &mut Vec<Selector<SelectorImpl>>,
-    non_common_style_affecting_attributes_selectors: &mut Vec<Selector<SelectorImpl>>)
-{
-    if stylesheet.disabled() || !stylesheet.is_effective_for_device(device) {
+fn add_stylesheet_internal(stylist: &mut Stylist, stylesheet: &Stylesheet) {
+    if stylesheet.disabled() || !stylesheet.is_effective_for_device(&stylist.device) {
         return;
     }
 
-    stylesheet.effective_rules(device, |rule| {
+    // Cheap `Arc` clone so that the closure below can borrow `&mut Stylist`.
+    let device = stylist.device.clone();
+
+    stylesheet.effective_rules(&device, |rule| {
         match *rule {
             CssRule::Style(ref style_rule) => {
                 let guard = style_rule.read();
                 for selector in &guard.selectors.0 {
                     let map = if let Some(ref pseudo) = selector.pseudo_element {
-                        pseudos_map
+                        stylist.pseudos_map
                             .entry(pseudo.clone())
                             .or_insert_with(PerPseudoElementSelectorMap::new)
                             .borrow_for_origin(&stylesheet.origin)
                     } else {
-                        element_map.borrow_for_origin(&stylesheet.origin)
+                        stylist.element_map.borrow_for_origin(&stylesheet.origin)
                     };
 
                     map.insert(Rule {
                         selector: selector.complex_selector.clone(),
                         style_rule: style_rule.clone(),
                         specificity: selector.specificity,
-                        source_order: *rules_source_order,
+                        source_order: stylist.rules_source_order,
                     });
                 }
-                *rules_source_order += 1;
+                stylist.rules_source_order += 1;
 
                 for selector in &guard.selectors.0 {
-                    state_deps.note_selector(&selector.complex_selector);
+                    stylist.state_deps.note_selector(&selector.complex_selector);
                     if selector.affects_siblings() {
-                        sibling_affecting_selectors.push(selector.clone());
+                        stylist.sibling_affecting_selectors.push(selector.clone());
                     }
 
                     if selector.matches_non_common_style_affecting_attribute() {
-                        non_common_style_affecting_attributes_selectors.push(selector.clone());
+                        stylist.non_common_style_affecting_attributes_selectors.push(selector.clone());
                     }
                 }
             }
             CssRule::Import(ref import) => {
                 let import = import.read();
-                add_stylesheet_internal(
-                    &import.stylesheet,
-                    device,
-                    element_map,
-                    pseudos_map,
-                    animations,
-                    precomputed_pseudo_element_decls,
-                    rules_source_order,
-                    state_deps,
-                    sibling_affecting_selectors,
-                    non_common_style_affecting_attributes_selectors)
+                add_stylesheet_internal(stylist, &import.stylesheet)
             }
             CssRule::Keyframes(ref keyframes_rule) => {
                 let keyframes_rule = keyframes_rule.read();
                 debug!("Found valid keyframes rule: {:?}", *keyframes_rule);
                 if let Some(animation) = KeyframesAnimation::from_keyframes(&keyframes_rule.keyframes) {
                     debug!("Found valid keyframe animation: {:?}", animation);
-                    animations.insert(keyframes_rule.name.clone(),
+                    stylist.animations.insert(keyframes_rule.name.clone(),
                                            animation);
                 } else {
                     // If there's a valid keyframes rule, even if it doesn't
                     // produce an animation, should shadow other animations
                     // with the same name.
-                    animations.remove(&keyframes_rule.name);
+                    stylist.animations.remove(&keyframes_rule.name);
                 }
             }
             // We don't care about any other rule.
@@ -188,7 +169,7 @@ impl Stylist {
     pub fn new(device: Device) -> Self {
         let mut stylist = Stylist {
             viewport_constraints: None,
-            device: device,
+            device: Arc::new(device),
             is_device_dirty: true,
             quirks_mode: false,
 
@@ -255,17 +236,7 @@ impl Stylist {
     }
 
     fn add_stylesheet(&mut self, stylesheet: &Stylesheet) {
-        add_stylesheet_internal(
-            stylesheet,
-            &self.device,
-            &mut self.element_map,
-            &mut self.pseudos_map,
-            &mut self.animations,
-            &mut self.precomputed_pseudo_element_decls,
-            &mut self.rules_source_order,
-            &mut self.state_deps,
-            &mut self.sibling_affecting_selectors,
-            &mut self.non_common_style_affecting_attributes_selectors);
+        add_stylesheet_internal(self, stylesheet);
 
         debug!("Stylist stats:");
         debug!(" - Got {} sibling-affecting selectors",
@@ -287,7 +258,6 @@ impl Stylist {
         })
     }
 
-
     /// Computes the style for a given "precomputed" pseudo-element, taking the
     /// universal rules and applying them.
     ///
@@ -419,7 +389,7 @@ impl Stylist {
             mq_eval_changed(&stylesheet.rules.read().0, &self.device, &device)
         });
 
-        self.device = device;
+        self.device = Arc::new(device);
     }
 
     pub fn viewport_constraints(&self) -> &Option<ViewportConstraints> {
```

---

*[components/style/stylist.rs, line 278 at r12](https://reviewable.io:443/reviews/servo/servo/14540#-KZ07ccFoLwEzJfJ-aCj:-KZ07ccGB3weB_EI4jq4:b-qpg3z2) ([raw file](https://github.com/servo/servo/blob/a62d681d978065e10f001c77c7b615b2672476a4/components/style/stylist.rs#L278)):*
> ```Rust
>                self.state_deps.len());
> 
>         SelectorImpl::each_precomputed_pseudo_element(|pseudo| {
> ```

This doesn’t need to run for every stylesheet, does it? Could we move it (and the `debug!` lines above) to the end of `Stylist::update`? That way the two `add_stylesheet*` methods/functions can be merged into one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14540)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:boom: Test timed out

emilio commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/script/stylesheet_loader.rs, line 151 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ03B-Fe6JkRo4L-hhV:-KZ0_AMgzgJJARH6uoOG:b-cs2jbr) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/script/stylesheet_loader.rs#L151)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

This should not panic. Instead, an invalid URL should produce a similar result as a network error.
</blockquote></details>

This doesn't panic, we don't send invalid urls to the loader altogether. This matches also what gecko does.

---

*[components/script/stylesheet_loader.rs, line 183 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ03SPfsnjxOQZ94Keu:-KZ0_HT01A_mGZn4xuAB:bruvgdl) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/script/stylesheet_loader.rs#L183)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

This should not panic. Instead, an invalid URL should produce a similar result as a network error.
</blockquote></details>

Same reasoning as above, urls that end up here aren't invalid.

---

*[components/style/media_queries.rs, line 20 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ04vsn4pSfIBu1A7XS:-KZ0_Um61itmeb579mV3:b-80tlqk) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/style/media_queries.rs#L20)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Please remove this `Clone` impl, see my "option dance" comment in `stylesheet_loader.rs`.
</blockquote></details>

Sure.

---

*[components/style/stylesheets.rs, line 48 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ04ju4Invm5zspxip8:-KZ0_XRBTU_9IbaYMOeb:b-i6ko4l) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/style/stylesheets.rs#L48)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Please revert this, it is never used.
</blockquote></details>

 Good catch.

---

*[components/style/stylist.rs, line 101 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ-qxdrCVWIfOI36wv7:-KZ0ZyQF7hrD9UmgNj9e:b-vs1pr4) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/style/stylist.rs#L101)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

This comment is not particularly helpful. I’d rather have it explain why this function takes some many arguments, but we can do better and fix that.

Please apply the diff below and move the function back to a `Stylist` method that takes `&mut self`. The new `Arc` does add some runtime cost, but I think it’s small enough that cleaner code is worth it.

```diff
diff --git a/components/style/stylist.rs b/components/style/stylist.rs
index 5b40dcf..5abef17 100644
--- a/components/style/stylist.rs
+++ b/components/style/stylist.rs
@@ -49,7 +49,7 @@ pub type FnvHashMap<K, V> = HashMap<K, V, BuildHasherDefault<::fnv::FnvHasher>>;
 #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
 pub struct Stylist {
     /// Device that the stylist is currently evaluating against.
-    pub device: Device,
+    pub device: Arc<Device>,
 
     /// Viewport constraints based on the current device.
     viewport_constraints: Option<ViewportConstraints>,
@@ -98,83 +98,64 @@ pub struct Stylist {
     non_common_style_affecting_attributes_selectors: Vec<Selector<SelectorImpl>>,
 }
 
-// I know...
-fn add_stylesheet_internal(
-    stylesheet: &Stylesheet,
-    device: &Device,
-    element_map: &mut PerPseudoElementSelectorMap,
-    pseudos_map: &mut FnvHashMap<PseudoElement, PerPseudoElementSelectorMap>,
-    animations: &mut FnvHashMap<Atom, KeyframesAnimation>,
-    precomputed_pseudo_element_decls: &mut FnvHashMap<PseudoElement, Vec<ApplicableDeclarationBlock>>,
-    rules_source_order: &mut usize,
-    state_deps: &mut DependencySet,
-    sibling_affecting_selectors: &mut Vec<Selector<SelectorImpl>>,
-    non_common_style_affecting_attributes_selectors: &mut Vec<Selector<SelectorImpl>>)
-{
-    if stylesheet.disabled() || !stylesheet.is_effective_for_device(device) {
+fn add_stylesheet_internal(stylist: &mut Stylist, stylesheet: &Stylesheet) {
+    if stylesheet.disabled() || !stylesheet.is_effective_for_device(&stylist.device) {
         return;
     }
 
-    stylesheet.effective_rules(device, |rule| {
+    // Cheap `Arc` clone so that the closure below can borrow `&mut Stylist`.
+    let device = stylist.device.clone();
+
+    stylesheet.effective_rules(&device, |rule| {
         match *rule {
             CssRule::Style(ref style_rule) => {
                 let guard = style_rule.read();
                 for selector in &guard.selectors.0 {
                     let map = if let Some(ref pseudo) = selector.pseudo_element {
-                        pseudos_map
+                        stylist.pseudos_map
                             .entry(pseudo.clone())
                             .or_insert_with(PerPseudoElementSelectorMap::new)
                             .borrow_for_origin(&stylesheet.origin)
                     } else {
-                        element_map.borrow_for_origin(&stylesheet.origin)
+                        stylist.element_map.borrow_for_origin(&stylesheet.origin)
                     };
 
                     map.insert(Rule {
                         selector: selector.complex_selector.clone(),
                         style_rule: style_rule.clone(),
                         specificity: selector.specificity,
-                        source_order: *rules_source_order,
+                        source_order: stylist.rules_source_order,
                     });
                 }
-                *rules_source_order += 1;
+                stylist.rules_source_order += 1;
 
                 for selector in &guard.selectors.0 {
-                    state_deps.note_selector(&selector.complex_selector);
+                    stylist.state_deps.note_selector(&selector.complex_selector);
                     if selector.affects_siblings() {
-                        sibling_affecting_selectors.push(selector.clone());
+                        stylist.sibling_affecting_selectors.push(selector.clone());
                     }
 
                     if selector.matches_non_common_style_affecting_attribute() {
-                        non_common_style_affecting_attributes_selectors.push(selector.clone());
+                        stylist.non_common_style_affecting_attributes_selectors.push(selector.clone());
                     }
                 }
             }
             CssRule::Import(ref import) => {
                 let import = import.read();
-                add_stylesheet_internal(
-                    &import.stylesheet,
-                    device,
-                    element_map,
-                    pseudos_map,
-                    animations,
-                    precomputed_pseudo_element_decls,
-                    rules_source_order,
-                    state_deps,
-                    sibling_affecting_selectors,
-                    non_common_style_affecting_attributes_selectors)
+                add_stylesheet_internal(stylist, &import.stylesheet)
             }
             CssRule::Keyframes(ref keyframes_rule) => {
                 let keyframes_rule = keyframes_rule.read();
                 debug!("Found valid keyframes rule: {:?}", *keyframes_rule);
                 if let Some(animation) = KeyframesAnimation::from_keyframes(&keyframes_rule.keyframes) {
                     debug!("Found valid keyframe animation: {:?}", animation);
-                    animations.insert(keyframes_rule.name.clone(),
+                    stylist.animations.insert(keyframes_rule.name.clone(),
                                            animation);
                 } else {
                     // If there's a valid keyframes rule, even if it doesn't
                     // produce an animation, should shadow other animations
                     // with the same name.
-                    animations.remove(&keyframes_rule.name);
+                    stylist.animations.remove(&keyframes_rule.name);
                 }
             }
             // We don't care about any other rule.
@@ -188,7 +169,7 @@ impl Stylist {
     pub fn new(device: Device) -> Self {
         let mut stylist = Stylist {
             viewport_constraints: None,
-            device: device,
+            device: Arc::new(device),
             is_device_dirty: true,
             quirks_mode: false,
 
@@ -255,17 +236,7 @@ impl Stylist {
     }
 
     fn add_stylesheet(&mut self, stylesheet: &Stylesheet) {
-        add_stylesheet_internal(
-            stylesheet,
-            &self.device,
-            &mut self.element_map,
-            &mut self.pseudos_map,
-            &mut self.animations,
-            &mut self.precomputed_pseudo_element_decls,
-            &mut self.rules_source_order,
-            &mut self.state_deps,
-            &mut self.sibling_affecting_selectors,
-            &mut self.non_common_style_affecting_attributes_selectors);
+        add_stylesheet_internal(self, stylesheet);
 
         debug!("Stylist stats:");
         debug!(" - Got {} sibling-affecting selectors",
@@ -287,7 +258,6 @@ impl Stylist {
         })
     }
 
-
     /// Computes the style for a given "precomputed" pseudo-element, taking the
     /// universal rules and applying them.
     ///
@@ -419,7 +389,7 @@ impl Stylist {
             mq_eval_changed(&stylesheet.rules.read().0, &self.device, &device)
         });
 
-        self.device = device;
+        self.device = Arc::new(device);
     }
 
     pub fn viewport_constraints(&self) -> &Option<ViewportConstraints> {
```

</blockquote></details>

Looks good.

---

*[components/style/stylist.rs, line 278 at r12](https://reviewable.io:443/reviews/servo/servo/14540#-KZ07ccFoLwEzJfJ-aCj:-KZ0Zu2qZ3JtB-fXzYdN:b2edl7j) ([raw file](https://github.com/servo/servo/blob/a62d681d978065e10f001c77c7b615b2672476a4/components/style/stylist.rs#L278)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

This doesn’t need to run for every stylesheet, does it? Could we move it (and the `debug!` lines above) to the end of `Stylist::update`? That way the two `add_stylesheet*` methods/functions can be merged into one.
</blockquote></details>

Whoops, right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14540)*
<!-- Sent from Reviewable.io -->

emilio commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/script/stylesheet_loader.rs, line 151 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ03B-Fe6JkRo4L-hhV:-KZ0bEfmbTgXSTjS8MH1:b-9qwljx) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/script/stylesheet_loader.rs#L151)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary><blockquote>

This doesn't panic, we don't send invalid urls to the loader altogether. This matches also what gecko does.
</blockquote></details>

Well, apparently I forgot to write that bit, I'll fix that too :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14540)*
<!-- Sent from Reviewable.io -->

emilio commented:
I've noticed that the spec actually specifies what should happen when imports fail to load, and that we also need to track pending loads from the style element, so I'm going to implement that properly too.

emilio commented:
@bors-servo try

Review comments addressed, load events cleaned up.

bors-servo commented:
:hourglass: Trying commit a578260 with merge d482de5...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1491)

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit a436800 with merge 9bc5256...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1458)

emilio commented:
So there are a few failing tests now, but I don't have a clear model to go forward, so I chose the behavior that made less test fail.

@Ms2ger: Previously we were firing both `error` and `load` at the same time. Some tests expect `error` being fired when the content-type is not `text/css`, something like we did very ad-hoc before in `process_response`. Other tests rely on `load` being fired with a non-`text/css` content-type. these are _a lot_ inside `"/_mozilla/mozilla/referrer-policy`.

Those passed before because we were firing both the `load` and `error` event, which I fixed. Those tests are a lot so I've kept firing the `load` event even if the content-type is not `text/css`. Do you agree? Should I revert that back and mark all those tests as failing instead?

emilio commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 16e05e5 with merge d3ed1ee...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4494), [arm32](http://build.servo.org/builders/arm32/builds/4669), [arm64](http://build.servo.org/builders/arm64/builds/4690), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4777), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1456), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1460), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4753), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4791), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1461), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1495), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2742)

emilio commented:
Ok, finally. r? @SimonSapin for the style system bits, @Ms2ger per my previous comment

SimonSapin commented:
Reviewed 5 of 5 files at r13, 4 of 6 files at r14.
Review status: 74 of 76 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/stylesheet_loader.rs, line 151 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ03B-Fe6JkRo4L-hhV:-KZ1wMQHFtO1sfR5P2Ig:b-y410p6) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/script/stylesheet_loader.rs#L151)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary><blockquote>

Well, apparently I forgot to write that bit, I'll fix that too :-)
</blockquote></details>

In that case, please change the `expect` message or add a comment to explain this and point to the code doing that (e.g. name a module and function)

---

*[components/script/stylesheet_loader.rs, line 183 at r8](https://reviewable.io:443/reviews/servo/servo/14540#-KZ03SPfsnjxOQZ94Keu:-KZ1wV3r_wf9WBd2esig:bkvr1g4) ([raw file](https://github.com/servo/servo/blob/a75f6a45c4a5bc54447268c6ba5f333c113ae798/components/script/stylesheet_loader.rs#L183)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary><blockquote>

Same reasoning as above, urls that end up here aren't invalid.
</blockquote></details>

Same reply as above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14540)*
<!-- Sent from Reviewable.io -->

emilio commented:
Done, also factored out those bits so I didn't have to write two comments :P

SimonSapin commented:
Ok, looks good to me except I’m unsure about the new test failures. Although this PR still progress overall so feel free to r=me or wait for input from @Ms2ger.

---

Reviewed 1 of 1 files at r15.
Review status: 74 of 76 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14540#-:-KZ36UiLGDa2y3eDLdyn:bbisjwd)*
<!-- Sent from Reviewable.io -->

Ms2ger commented:
Feel free to land if I haven't managed to review by EOD Friday.

emilio committed:
style: Add a struct to represent import rules, and parse them correctly.

emilio committed:
style: Add a StylesheetLoader abstraction, and make it a no-op on Geckolib.

Servo doesn't compile at this stage.

emilio committed:
style: Add a way to update an empty stylesheet for @import.

We'll update the empty stylesheet we've created when instantiating the
ImportRule when the stylesheet finishes loading.

emilio committed:
script: Add CSSImportRule WebIDL interface.

emilio committed:
script: Add infrastructure to track subresource loads in <link> and <style> elements.

emilio committed:
script: Move the stylesheet loading code from htmllinkelement to stylesheet_loader. Make it track subresource loads properly.

emilio committed:
Update test expectations.

Ms2ger commented on components/script/stylesheet_loader.rs:
```diff
@@ -0,0 +1,254 @@
+            if link.parser_inserted() {
+                document.increment_script_blocking_stylesheet_count();
+            }
+            if link.RelList().Contains("noreferrer".into()) {
```
Let's pass this in as an argument, and have a trait that `HTMLLinkElement` and `HTMLStyleElement` both implement, instead of duplicating all the code.

emilio replied:
Yeah, I'll make a trait for both, and add some `override_referrer_policy(&self) -> Option<ReferrerPolicy>` or similar.

emilio commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -217,6 +234,28 @@ impl VirtualMethods for HTMLLinkElement {
+        }
+
+        let any_failed = self.any_failed_load.get();
+        self.any_failed_load.set(false);
```
Because at that point you ran out of loads, and you don't want the next load that may be triggered by changing the `href` to think it's failed unconditionally.

emilio commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -217,6 +218,15 @@ impl VirtualMethods for HTMLLinkElement {
+    }
+
+    pub fn decrement_pending_loads_count(&self) -> bool {
+        self.pending_loads.set(self.pending_loads.get() - 1);
```
I think this is unused now, I switched to use `load_finished`, so I'll just get rid of it.

emilio replied:
(Indeed it was removed later).

emilio commented on components/script/dom/htmllinkelement.rs:
```diff
@@ -1,12 +1,10 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
+* License, v. 2.0. If a copy of the MPL was not distributed with this
```
It was reverted in a later commit.

emilio committed:
script: Clarify htmllinkelement's comment about `pending_loads`.

SimonSapin commented on components/style/stylesheets.rs:
```diff
@@ -345,6 +359,34 @@ impl ToCss for NamespaceRule {
+            try!(dest.write_str(" "));
+            try!(media.to_css(dest));
+        }
+        Ok(())
```
Yes it is, sorry I missed this. +1 on adding a test.

emilio committed:
script: Abstract HTMLLinkElement and StyleElement into StylesheetOwner.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14615) made this pull request unmergeable. Please resolve the merge conflicts.

emilio committed:
style: Fix @import serialization.

emilio commented:
Rebased, comments hopefully addressed, re-r? @Ms2ger

SimonSapin commented:
New test looks good.

---

Reviewed 1 of 6 files at r25, 53 of 53 files at r28, 2 of 3 files at r31.
Review status: 56 of 78 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14540#-:-KZ7ZoIWVZ7aISoVpUKe:b-ucuo5g)*
<!-- Sent from Reviewable.io -->

emilio commented:
@bors-servo r=SimonSapin,Ms2ger

http://logs.glob.uno/?c=mozilla%23servo&s=16+Dec+2016&e=16+Dec+2016#c578565

bors-servo commented:
:pushpin: Commit 8356c33 has been approved by `SimonSapin,Ms2ger`

<!-- @bors-servo r=SimonSapin,Ms2ger 8356c3386abdcdb16bfbf9377908b0d24d3dc588 -->

bors-servo commented:
:hourglass: Testing commit 8356c33 with merge 38f1361...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4515), [arm32](http://build.servo.org/builders/arm32/builds/4689), [arm64](http://build.servo.org/builders/arm64/builds/4711), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4798), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1478), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1481), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4774), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4812), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1482), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1516), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2763)
